scipy/pixi.toml

```diff
@@ -152,7 +152,7 @@ test-cupy = { cmd = "python dev.py test -b cupy", cwd = "scipy", env = { SCIPY_A
 platforms = ["linux-64", "osx-arm64"]
 
 [feature.jax.dependencies]
-jax = "=0.4.28"
+jax = "*"
```

Comment:
I am afraid that this may install a cpu-only jax if there is a conflicting dependency (like a new jax version available, but not available for the cuda+cudnn version that pytorch requires, as it happens now were the latest jax requires cudnn==9 and there is no pytorch build with cuda compatible with it). I guess you can add `jaxlib = {version = "*", build = "*cuda*"}` to make sure that cuda-enabled jax gets installed, see https://github.com/ami-iit/jaxsim/pull/221#discussion_r1724820020 for a related discussion.

Reply:
That by itself can't be the needed change (if a change is needed), because we also have environments where we want cpu-only `jax`. There are `cpu` and `cuda` features that follow the example in https://pixi.sh/latest/features/multi_environment/#real-world-example-use-cases. The environment definition then combines the features like:
```
env-name = ["cuda", "jax", "pytorch"]
```
If CUDA 12 is present in the environment, the pytorch/jax CUDA versions should be chosen. If that doesn't happen then there's indeed a problem, but I haven't seen that yet. 

Do you happen to know if `jax`/`jaxlib` in conda-forge uses the same mutex as `pytorch` for selection the CPU or CUDA build flavor?

Reply:
> That by itself can't be the needed change (if a change is needed), because we also have environments where we want cpu-only jax. 

My bad, I did not checked the full `pixi.toml`. In that case the constraint needs to be added in an appropriate cuda-related feature I guess.

> Do you happen to know if jax/jaxlib in conda-forge uses the same mutex as pytorch for selection the CPU or CUDA build flavor?

I am afraid there is no mutex package for pytorch (in the sense of `pytorch-mutex`) on conda-forge. I guess for both jaxlib and pytorch the cuda build strings starts with `cuda`.

Reply:
Ah:( IIRC there was a mutex years ago, but https://github.com/conda-forge/pytorch-cpu-feedstock/pull/247 discusses the current state and it's quite messy.

Adding the build string really doesn't compose well. I'll look into this more. What I've seen so far is that it works as advertised; probably `pixi` is relying on `system-requirements = { cuda = "12" }` to gate selecting `cuda` build variants on. I haven't seen it pick a `cpu` build yet with that present, however it's not clear to me whether cuda/cpu is retained when there's a version conflict. Will do some more in-depth testing.